### ISSUE TITLE
sb_param_estim__copasi pipeline executed on sge or lsf cluster

### ISSUE THREAD

**Issue by pdp10** [pdp10/sb_pipe#46]
If sb_pipe is installed on cluster with support for either sge or lsf, CopasiSE should be executed as cluster jobs. 

Bash scripts for this are in: https://github.com/wikiselev/coparaest
@wikiselev
If interested, would you like to merge that code in sb_pipe so that our efforts are collected in one tool? 

This requires the inclusion of a new option in the configuration file for parameter estimation: 
e.g. 
cluster=sge || lsf || pp

- If pp is used, the pipeline works as already implemented using parallel python. 
- If sge or lsf are used, the pipeline would start the jobs with bsub or qsub. This would be done using Python subprocess.Popen instead of Bash. This selection could be made in the pipeline main script: sb_param_estim__copasi.py (maybe).

Some of the plots could also be merged. The selection of the best fit via % instead of the selection of the 10 best fits should be maintained though. I think all those plots can be created from the summary table generated by sb_pipe. This would exclude the use of the Perl script, simplifying the work and removing a dependency.

**Comment by wikiselev** [comment]
Hi Piero, could you please give me a step by step instructions for the pull request 👍  
I already forgot how coparaest works 👍

**Comment by pdp10** [comment]
Hi Vladimir, I think you just need to fork it and then after editing you can request a pull request from your code into this one. 
Alternatively, I can make you a contributor, so that you don't need to fork the project as you can commit directly. To commit in develop, the only way is via pull request (I decided this for safety mainly). So you should make your commits on a branch from develop and then open a pull request to merge your branch into develop.

**Comment by pdp10** [comment]
The `develop` branch now contains an updated structure for the script: 
`https://github.com/pdp10/sb_pipe/blob/master/sb_pipe/sb_param_estim__copasi.py` .

In particular: 

1. the new configuration file for this pipeline contains a new variable `cluster` which can be pp, sge or lsf. 
2. The script now contains the below code which needs to be edited for lsf and sge. 
```
  if cluster == "lsf":
    for i in xrange(1:nfits):
      # TODO Correct the following line
      process = subprocess.Popen(["bsub", "-J", "CopasiSE", "-s", model[:-4]+str(i)+".cps", model[:-4]+str(i)+".cps"])
    # TODO Check here when these jobs are finished before proceeding
  elif cluster == "sge":
    for i in xrange(1:nfits):
      # TODO Correct the following line      
      process = subprocess.Popen(["qsub", "CopasiSE", "-s", model[:-4]+str(i)+".cps", model[:-4]+str(i)+".cps"])`
    # TODO Check here when these jobs are finished before proceeding
```

**Comment by wikiselev** [comment]
Hi Pierro, had no time to deal with this, but looks like you've done everything yourself! 👍

**Comment by pdp10** [comment]
Hi Vladimir, 
Yes, I need to run some estimations and although my model is small, this comes out nicely. 
For a LSF cluster, I don't think it is completely correct though. I suspect the pipeline continues without waiting for the processes to terminate. Although there is an option for this using bsub, the problem is that I am starting these from Python and so, I would need an option in bsub which "locks" the python subprocess Popen() call. 

Aside from this.. I have a weird problem with R coming out AFTER the parameter estimation. It seems related to ggplot2.. See here: https://github.com/pdp10/sb_pipe/issues/58
It only happens when the pipeline is run on the cluster, not on console. 

Any thought?

**Comment by pdp10** [comment]
Okay, problem solved for LSF.
The problem related to R ggplot2 using a cluster remains instead. This is addressed on Issue #58 .  
Current issue - closed.
